dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java

```diff
@@ -0,0 +1,370 @@
+        OrganisationUnit ou = organisationUnitService.getOrganisationUnit(ouid);
+        User user = userService.getUser(userid);
+
+        if (ou == null) {
+        	throw new SMSProcessingException(SMSResponse.INVALID_ORGUNIT.set(ouid));
+        } else if (user == null) {
+        	throw new SMSProcessingException(SMSResponse.INVALID_USER.set(userid));
+        } else if (!user.getOrganisationUnits().contains(ou)) {
+        	throw new SMSProcessingException(SMSResponse.USER_NOTIN_OU.set(userid, ouid));
+        }
+	}
+	
+	private SMSSubmissionHeader getHeader( IncomingSms sms ) {
+		byte[] smsBytes = SmsUtils.getBytes(sms);
+		SMSSubmissionReader reader = new SMSSubmissionReader();
+		try {
+			return reader.readHeader(smsBytes);
+		} catch ( Exception e ) {
+			Log.error(e.getMessage());
+			e.printStackTrace();
+			return null;
+		}
+	}
+
+	public SMSMetadata getMetadata( Date lastSyncDate ) {
+		SMSMetadata meta = new SMSMetadata();
+		meta.dataElements = getAllDataElements(lastSyncDate);
+		meta.categoryOptionCombos = getAllCatOptionCombos(lastSyncDate);
+		meta.organisationUnits = getAllUserIds(lastSyncDate);
+		meta.trackedEntityTypes = getAllTrackedEntityTypeIds(lastSyncDate);
+		meta.trackedEntityAttributes = getAllTrackedEntityAttributeIds(lastSyncDate);
+		meta.programs = getAllProgramIds(lastSyncDate);
+		meta.organisationUnits = getAllOrgUnitIds(lastSyncDate);
+
+		return meta;
+	}
+
+    public List<SMSMetadata.ID> getAllUserIds (Date lastSyncDate) {
+        List<User> users = userService.getAllUsers();
+
+        return users
+        		.stream()
+        		.map(o -> getIdFromMetadata(o, lastSyncDate))
+				.filter(Objects::nonNull)
+        		.collect(Collectors.toList());
+    }
+
+    public List<SMSMetadata.ID> getAllTrackedEntityTypeIds (Date lastSyncDate) {
+        List<TrackedEntityType> teTypes = trackedEntityTypeService.getAllTrackedEntityType();
+
+        return teTypes
+        		.stream()
+        		.map(o -> getIdFromMetadata(o, lastSyncDate))
+				.filter(Objects::nonNull)
+        		.collect(Collectors.toList());
+    }
+
+    public List<SMSMetadata.ID> getAllTrackedEntityAttributeIds (Date lastSyncDate) {
+        List<TrackedEntityAttribute> teiAttributes = trackedEntityAttributeService.getAllTrackedEntityAttributes();
+
+        return teiAttributes
+        		.stream()
+        		.map(o -> getIdFromMetadata(o, lastSyncDate))
+				.filter(Objects::nonNull)
+        		.collect(Collectors.toList());
+    }
+
+    public List<SMSMetadata.ID> getAllProgramIds (Date lastSyncDate) {
+        List<Program> programs = programService.getAllPrograms();
+
+        return programs
+        		.stream()
+        		.map(o -> getIdFromMetadata(o, lastSyncDate))
+				.filter(Objects::nonNull)
+        		.collect(Collectors.toList());
+    }
+
+    public List<SMSMetadata.ID> getAllOrgUnitIds (Date lastSyncDate) {
+        List<OrganisationUnit> orgUnits = organisationUnitService.getAllOrganisationUnits();
```

Comment:
these are heavy operations. Wouldn't it be good if we filter at SQL level?

Reply:
Yeah, that's a good point, I suppose if we just need the IDs we shouldn't run a query to get all the metadata.

Could you point me to an example in the code where something like this has been done before? As in we would replace it with a hard coded SQL query?